### ISSUE TITLE
Speed accessing single resource

### ISSUE THREAD

**Issue by devinrsmith** [classgraph/classgraph#378]
Hi,

I'm really impressed by classgraph - classloading is tricky, and I'm glad to see a library like this make it easier :).

Is accessing a single non-class resource via classgraph an intended use-case? It's clearly possible, but I'm finding that it is about an order of magnitude slower than loading directly from a `ClassLoader` (on a variety of classpath sizes), and startup speed is in an important concern for my current situation. I've tried a variety of whitelisting, and disabling different types of scanning, but I'm not seeing the speed that I'm hoping to achieve.

```java
  private void processResourceClassloader(ClassLoader loader, String name) throws IOException {
    final Enumeration<URL> urls = loader.getResources(name);
    while (urls.hasMoreElements()) {
      final URL url = urls.nextElement();
      try (final InputStream in = url.openStream()) {
        doFastOperation(in);
      }
    }
  }

  private void processResourceClassgraph(ClassLoader loader, String name) throws IOException {
    try (final ScanResult scan = new ClassGraph()
        .overrideClassLoaders(loader)
        .disableNestedJarScanning()
        .disableModuleScanning()
        .whitelistClasspathElementsContainingResourcePath(name)
        //.whitelistPaths(name)
        .scan()) {
      for (final Resource resource : scan.getResourcesWithPath(name)) {
        try (final InputStream in = resource.open()) {
          doFastOperation(in);
        }
      }
    }
  }
```

**Comment by lukehutch** [comment]
@devinrsmith Thanks, I'm glad it has worked well for you (other than the latency issue), it took a ton of work to get ClassGraph to where it is today.

The startup latency is indeed an issue. I have profiled the heck out of scanning large classpaths, to ensure maximum scanning speed and maximum parallelization. But I never spent the time to deeply profile and understand how startup speed could be improved. And it's definitely worth improving.

Do you have experience with profiling JVM applications? If so, I would appreciate your help in figuring out where the time is taken right now. If not, I'll see if I can get to it sometime soon, although I have a lot on my plate right now.

**Comment by lukehutch** [comment]
BTW comparing the speed of ClassGraph to getting the resource through the current classloader is not an apples-to-apples comparison, because the current classloader has already incurred a one-time startup cost to open all jars on the classpath and read (at a minimum) their zip directory entries, which are cached. ClassGraph has to repeat all of this work, because at most it can only obtain the URIs of the classpath from the classloader (which it has to do with reflection in most cases) -- it still has to open and read all jars and/or all directories on the classpath, once the paths have been obtained. So you may not be able to bring the performance as close as you would like.

I notice you are calling `overrideClassLoaders(loader)`, which will help. But since you're already calling `.whitelistClasspathElementsContainingResourcePath(name)`, maybe you can create your own custom `URLClassLoader` containing only URLs of jars that can possibly contain the resource you're looking for (as long as you can determine this a priori). That will reduce the startup time, since then ClassGraph doesn't have to pre-scan all classpath entries (opening all jarfiles to read their directories, and checking all jars and dirs to see if they contain the requested path) before attempting to read that resource from the matching classpath elements.

**Comment by devinrsmith** [comment]
I've done a bunch of profiling / JVM optimizations over the years. [JMH](https://openjdk.java.net/projects/code-tools/jmh/) is a good tool for micro-benchmarking, and probably the best place to start, but it might not be the best for capturing startup speed behavior. Trying to profile application startup (or cold-path code) is a tougher problem - there are ways to attach JVM agents (https://github.com/jvm-profiling-tools/perf-map-agent is one example), or the java flight recorder, to get out timing data, which can be analyzed using a number of different methods (flame graphs are popular, but I don't have too much experience w/ that style). 

I'm happy to help dig in as timing permits. Cheers!

**Comment by lukehutch** [comment]
It's detailed timing traces that are needed, more than microbenchmarks. JMH comes in handy once you're trying to figure out the total wall-clock time impact of a given optimization.

With something as short lived as this, it's generally better to start with repeating the same code path a large number of times, and run the code through a profiler to identify hotspots. This doesn't test cold-start overhead, but at least it identifies where the time is being spent once the system is warmed up.

**Comment by lukehutch** [comment]
Just a thought on benchmarking -- really the code you need to benchmark against is code that opens a `ZipFile` instance on the jar of interest, and iterates through all the `ZipEntry` entries in the jar. If you're including the time to actually load the content of the resource of interest in the benchmark, then you'll also need to open an `InputStream` on the appropriate `ZipEntry` when you get to it, and read the entire stream. ClassGraph has to do all that work to scan the jar and read the resource of interest, so that will give a lower time bound to compare against.

Actually ClassGraph implements its own zipfile library for speed and flexibility (it is particularly fast compared to Java's API when reading nested jars, since Java has to extract nested jars to temporary files to work with them, and ClassGraph just works with file slices for speed). So I would be surprised if ClassGraph adds much overhead compared to the `ZipFile` code described above, unless you are scanning many more jars than you need to in the classpath.

**Comment by devinrsmith** [comment]
I've put together a quick benchmark runner for startup speed loading resources from jars that are explicitly not part of the initial jvm classpath.

https://github.com/devinrsmith/classgraph-benchmark

Here is an example output https://gist.github.com/devinrsmith/dd899dd5c515ab90a985b58a8677b799

A snippet from that:
```
classloader META-INF/maven/com.google.guava/guava/pom.xml 2 84a3ee12 17 56
real 0.10
user 0.13
sys 0.03
classgraph META-INF/maven/com.google.guava/guava/pom.xml 2 84a3ee12 122 162
real 0.22
user 0.60
sys 0.05
```

Essentially, the classloader took 17 millis (56 millis from main) to load and hash META-INF/maven/com.google.guava/guava/pom.xml; and classgraph took 122 millis (162 millis from main) to load and hash the same resources.

**Comment by lukehutch** [comment]
Good work -- yeah, an order of magnitude difference in time is really not acceptable! So in both cases, you defined a custom `URLClassLoader` with a single jar URL, and used that to get the resource?

If you run each of these tests 1000x in a loop, a profiler should be able to identify hotspots. But also, take a look at the verbose log output of ClassGraph for a single run through, to see if you can see anywhere it is doing unnecessary work. Thanks!

**Comment by lukehutch** [comment]
I did some extensive profiling of ClassGraph, and found a number of opportunities for optimization, mostly in how zipfile entries were being parsed (e.g. it was possible to defer parsing of modification date and permission information until the user requests that info for a `Resource`). ClassGraph is a bit faster now. (Released in 4.8.55.)

However, unfortunately I don't see any other major avenues for optimization. ClassGraph has to do additional work on startup that a regular classloader does not, for example starting up a thread pool, querying all classloaders using reflection to find classpath entries, and reading any directories, jarfile central directory entries, and modules on the classpath and module path for every scan performed (whereas classloaders can cache this information).

I tried to create an apples-to-apples comparison, by copying a jarfile 500x on disk (to prevent a Java classloader from cheating by caching the jarfile central directory info somewhere, either in the JRE, or the classloader, or the classloader's parent). ClassGraph was called with `overrideClassLoaders(classLoader)` for a single target classLoader, which disables scanning of modules and other classloaders.

The results indicate that ClassGraph 4.8.55 is about 4.5x slower than `URLClassLoader` at loading a single resource (this varies between 3.5-7x for large jars). I suspect the performance gap would shrink as you load more and more resources from the same jars, since the cost of reading the jarfile central directory can be amortized across all the resources read.

Right now the biggest bottleneck in ClassGraph is a method that reads an unsigned short from a `byte[]` array using `((arr[i + 1] & 0xff) << 8) | (arr[i] & 0xff)`, in order to read the jarfile central directory header for the jar. This takes almost 30% of the total scan time, and collectively all the central directory parsing code takes about half the total scan time. There's no way to speed this up other than rewriting in native code, unfortunately. The JRE's own `ZipFile.getEntry()` does this work in native code, which is one of the main speed advantages. Unfortunately the JRE method can't be used, because the `ZipFile` API uses a mutex around all API calls, and ClassGraph is parallelized.

So I think this comes down to a question of whether you value ClassGraph's flexibility (its ability to work with a [wide range of different classloaders and classpath specification mechanisms](https://github.com/classgraph/classgraph/wiki/Classpath-Specification-Mechanisms)), or the JRE's own speed advantage, obtained by calling into native code.

However, if you are performing more than one scan, you can at least save some startup time in ClassGraph by providing your own `ExecutorService` to the `scan` method. (Don't forget to shut down the `ExecutorService` when you have finished with it.) More than 1 thread probably won't help if you're pulling a single resource from a single jarfile, but if you're pulling multiple resources from multiple jarfiles, or additionally scanning classfiles, more threads should help. You can see this usage pattern in the benchmark code below.

```java
import java.io.File;
import java.io.IOException;
import java.net.URL;
import java.net.URLClassLoader;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.ThreadFactory;

import io.github.classgraph.ClassGraph;
import io.github.classgraph.ScanResult;
import nonapi.io.github.classgraph.utils.FileUtils;

public class ClassGraphStartupBenchmarkMain {

    private static void createTempFiles(String jarPath, List<String> paths) throws Exception {
        for (String path : paths) {
            final var cmd = "cp -f " + jarPath + " " + path;
            int result = Runtime.getRuntime().exec(cmd).waitFor();
            if (result != 0) {
                throw new IOException("Command returned status code " + result + ": " + cmd);
            }
        }
    }

    private static void deleteTempFiles(List<String> paths) {
        for (String path : paths) {
            new File(path).delete();
        }
    }

    public static void main(String[] args) throws Exception {
        // 32MB, 16870 entries:
        String jarPath = "/home/luke/.m2/repository/org/jetbrains/kotlin/kotlin-compiler/1.3.21/kotlin-compiler-1.3.21.jar";
        // 160kB:
        String resourcePath = "win32/amd64/liblz4-java.so";

        //        // 472kB, 245 entries:
        //        String jarPath = "/home/luke/.m2/repository/io/github/classgraph/classgraph/4.8.54/classgraph-4.8.54.jar";
        //        // 2.8kB:
        //        String resourcePath = "META-INF/MANIFEST.MF";

        String tmpFilePathPrefix = "/home/luke/Downloads/";
        int N = 500;

        List<String> paths = new ArrayList<>();
        for (int i = 0; i < N; i++) {
            final var path = tmpFilePathPrefix + "benchmark-tmp-" + i + ".jar";
            paths.add(path);
        }

        try {
            System.out.println("Creating temp files");
            createTempFiles(jarPath, paths);

            // ClassGraph: -----------------------------------------------------------------------------------------

            System.out.println("Testing ClassGraph speed");
            ExecutorService executorService = null;
            try {
                // Set up ExecutorService so that ClassGraph doesn't have to start a new one for each new scan
                int numWorkerThreads = 1;
                executorService = Executors.newFixedThreadPool(numWorkerThreads, new ThreadFactory() {
                    public Thread newThread(Runnable r) {
                        Thread t = new Thread(r);
                        // Kill worker threads if main thread dies
                        t.setDaemon(true);
                        return t;
                    }
                });

                // Recreate classloaders so caching does not affect timing
                List<URLClassLoader> classLoadersC = new ArrayList<>();
                for (String path : paths) {
                    classLoadersC.add(new URLClassLoader(new URL[] { new URL("file://" + path) }));
                }

                long t2 = System.nanoTime();
                for (URLClassLoader classLoader : classLoadersC) {
                    try (ScanResult scanResult = new ClassGraph().overrideClassLoaders(classLoader)
                            .scan(executorService, numWorkerThreads)) {
                        scanResult.getResourcesWithPath(resourcePath).forEachByteArray((resource, byteArray) -> {
                            // Resource content has been read
                        });
                    }
                }
                long t3 = System.nanoTime();
                
                System.out.printf("ClassGraph took %.3f sec\n", (t3 - t2) * 1.0e-9);

                for (URLClassLoader classLoader : classLoadersC) {
                    classLoader.close();
                }

            } finally {
                if (executorService != null) {
                    executorService.shutdownNow();
                }
            }

            // Recreate temp files so there's no caching advantage -------------------------------------------------
            
            System.out.println("Recreating temp files");
            deleteTempFiles(paths);
            createTempFiles(jarPath, paths);

            // URLClassLoader: -------------------------------------------------------------------------------------

            System.out.println("Testing URLClassLoader speed");
            long t0 = System.nanoTime();
            List<URLClassLoader> classLoadersU = new ArrayList<>();
            for (String path : paths) {
                classLoadersU.add(new URLClassLoader(new URL[] { new URL("file://" + path) }));
            }
            for (URLClassLoader classLoader : classLoadersU) {
                URL resURL = classLoader.findResource(resourcePath);
                try (var is = resURL.openStream()) {
                    FileUtils.readAllBytesAsArray(is, -1);
                }
                classLoader.close();
            }
            long t1 = System.nanoTime();

            System.out.printf("URLClassLoader took %.3f sec\n", (t1 - t0) * 1.0e-9);

        } finally {
            // Delete temp files
            System.out.println("Deleting temp files");
            deleteTempFiles(paths);
        }
        System.out.println("Finished");
    }

    // Times for ClassGraph / URLClassLoader (multiple runs)
    //
    //     N=500, 32MB jar, 16870 entries, reading 160kB resource:
    //
    //         9.942 / 1.642 = 6.05x slower
    //         13.345 / 3.727 = 3.58x slower
    //         11.680 / 1.628 = 7.17x slower 
    //         10.457 / 2.545 = 4.11x slower 
    //         11.146 / 2.497 = 4.46x slower
    //         10.430 / 2.553 = 4.09x slower 
    //
    //     N=500, 472kB jar, 245 entries, reading 2.8kB resource
    //
    //         0.729 / 0.158 = 4.61x slower
    //         0.794 / 0.168 = 4.73x slower
    //         0.776 / 0.158 = 4.91x slower
    //         0.779 / 0.157 = 4.96x slower
}
```

Unfortunately I think that's about all that can be done for now unless native code is written to accelerate zipfile central directory parsing. Sorry about that!

You're welcome to profile this further though if you want to see if anything else can be sped up for your specific usecase, and I'll reopen this if you find more things that can be optimized.
